### ISSUE TITLE
ULMAT numerical instability in stacked meshes

### ISSUE THREAD

**Issue by marcosvanella** [firemodels/fds#11307]
This is a 5 mesh case, the top mesh is stretched in z. It runs with default solver but gives numerical instability with ULMAT, even though the FishPack solver should be called in all meshes.

```
&HEAD CHID='test_ULMAT', TITLE='ULMAT test.' /

# meshes total:
&MESH IJK=8,8,32, XB=-1400,-600,-1800,-1000,236,1000, TRNZ_ID='MY TRANSFORM', COLOR='BLUE', CHECK_MESH_ALIGNMENT=T / 1 mesh 4 m vertical resolution at interface
&MULT ID='m1', DX=400.00, DY=400.00,
      I_LOWER=-3,I_UPPER=-2,
      J_LOWER=-4,J_UPPER=-3/
&MESH IJK=16,16,16, XB=-200.00,200.00,-200.00,200.00,-20,236,  MULT_ID='m1', COLOR='GRAY', CHECK_MESH_ALIGNMENT=T/

&TIME T_END=1200/

#&PRES SOLVER='ULMAT'/ 

&RADI RADIATION=F /

&WIND INITIAL_SPEED=10.0, DIRECTION=206 /

! External BCs:
&VENT MB='XMIN',SURF_ID='OPEN'/
&VENT MB='XMAX',SURF_ID='OPEN'/
&VENT MB='YMIN',SURF_ID='OPEN'/
&VENT MB='YMAX',SURF_ID='OPEN'/
&VENT MB='ZMAX',SURF_ID='OPEN'/


&TRNZ ID='MY TRANSFORM', CC= 283.750000000000000, PC= 268.000000000000000 /
&TRNZ ID='MY TRANSFORM', CC= 307.625000000000000, PC= 284.000000000000000 /
&TRNZ ID='MY TRANSFORM', CC= 331.500000000000000, PC= 300.000000000000000 /
&TRNZ ID='MY TRANSFORM', CC= 355.375000000000000, PC= 316.000000000000000 /
&TRNZ ID='MY TRANSFORM', CC= 379.250000000000000, PC= 332.000000000000000 /
&TRNZ ID='MY TRANSFORM', CC= 403.125000000000000, PC= 348.000000000000000 /
&TRNZ ID='MY TRANSFORM', CC= 427.000000000000000, PC= 364.000000000000000 /
&TRNZ ID='MY TRANSFORM', CC= 450.875000000000000, PC= 380.000000000000000 /
&TRNZ ID='MY TRANSFORM', CC= 474.750000000000000, PC= 396.000000000000000 /
&TRNZ ID='MY TRANSFORM', CC= 498.625000000000000, PC= 412.000000000000000 /
&TRNZ ID='MY TRANSFORM', CC= 522.500000000000000, PC= 428.000000000000000 /
&TRNZ ID='MY TRANSFORM', CC= 546.375000000000000, PC= 444.000000000000000 /
&TRNZ ID='MY TRANSFORM', CC= 570.250000000000000, PC= 460.000000000000000 /
&TRNZ ID='MY TRANSFORM', CC= 594.125000000000000, PC= 476.000000000000000 /
&TRNZ ID='MY TRANSFORM', CC= 618.000000000000000, PC= 492.000000000000000 /
&TRNZ ID='MY TRANSFORM', CC= 641.875000000000000, PC= 509.226830095770538 /
&TRNZ ID='MY TRANSFORM', CC= 665.750000000000000, PC= 527.774559792554669 /
&TRNZ ID='MY TRANSFORM', CC= 689.625000000000000, PC= 547.744471551853280 /
&TRNZ ID='MY TRANSFORM', CC= 713.500000000000000, PC= 569.245613858413776 /
&TRNZ ID='MY TRANSFORM', CC= 737.375000000000000, PC= 592.395396694670126 /
&TRNZ ID='MY TRANSFORM', CC= 761.250000000000000, PC= 617.320232674305885 /
&TRNZ ID='MY TRANSFORM', CC= 785.125000000000000, PC= 644.156227335939320 /
&TRNZ ID='MY TRANSFORM', CC= 809.000000000000000, PC= 673.049922366374631 /
&TRNZ ID='MY TRANSFORM', CC= 832.875000000000000, PC= 704.159095811894531 /
&TRNZ ID='MY TRANSFORM', CC= 856.750000000000000, PC= 737.653623647258769 /
&TRNZ ID='MY TRANSFORM', CC= 880.625000000000000, PC= 773.716407407126098 /
&TRNZ ID='MY TRANSFORM', CC= 904.500000000000000, PC= 812.544372945360351 /
&TRNZ ID='MY TRANSFORM', CC= 928.375000000000000, PC= 854.349545776085051 /
&TRNZ ID='MY TRANSFORM', CC= 952.250000000000000, PC= 899.360208868536120 /
&TRNZ ID='MY TRANSFORM', CC= 976.125000000000000, PC= 947.822149218012783 /

&TAIL/
```

**Comment by rmcdermo** [comment]
Do `git show 1e8f6ca03`

**Comment by marcosvanella** [comment]
That looks correct at first glance. Do a grep on the source in IPS. It will drive what FFT Poisson solver to use. Depending on where the stretching is happening, source, bcs and solution arrays can get transposed. We don't want to do that, and keep arrays such that x,y,z -> I,J,K regardless of stretching.

**Comment by marcosvanella** [comment]
I think the issue here is that for the FFT mesh (i.e. stretched at top) this was changed to IPS=0 when IPS should be whatever it was computed. IPS should depend on ZM%USE_FFT for ULMAT.

**Comment by rmcdermo** [comment]
The stacked mesh issue is resolved with PR #11329 .  However, if a GEOM is added with a stretch mesh, the case fails.

[geom_stretched_grid.fds.txt](https://github.com/firemodels/fds/files/10403657/geom_stretched_grid.fds.txt)

**Comment by rmcdermo** [comment]
GEOM case seems to be fixed with PR #11330 .  Leave open until we get a verification case established.

**Comment by rmcdermo** [comment]
Firebot seems happy with the fix.
